synapse/api/errors.py

```diff
@@ -600,8 +606,12 @@ def cs_error(msg: str, code: str = Codes.UNKNOWN, **kwargs: Any) -> "JsonDict":
 
 
 class FederationError(RuntimeError):
-    """This class is used to inform remote homeservers about erroneous
-    PDUs they sent us.
+    """
+    This class is used signal about erroneous PDUs that we pulled from a remote homeserver.
```

Comment:
Because there appears to be two different use cases for `FederationError`, it seems like we should have two different error types; one for processing pulled events, and another to respond with when remote homeserver pushes events to us via `/send`.

We can tackle this in a future PR. And if others agree, I can add a `TODO: note` in this docstring here

Reply:
> We can tackle this in a future PR. And if others agree, I can add a TODO: note in the docstring here

Sounds sensible.

----

`FederationError` isn't a great name. `InvalidPDUError` maybe? But it might be a pain to rename now, and it sounds like you have plans to break this down into finer exception classes.

I wasn't 100% following your phrasing. Here's another suggestion representing my tenuous understanding:

```
Raised when we process an erroneous PDU.

There are two kinds scenarios where this exception can be raised:

1. We may pull an invalid PDU from a remote homeserver (e.g. during backfill).
   We raise this exception to signal an error to the rest of the application.
2. We may be pushed an invalid PDU as part of a `/send` transaction from
   a remote homeserver. We raise so that we can respond to the transaction with
   an (unspecced) error response and status code, which will likely be ignored.
```

Reply:
Your new language sounds good to me 👍  Updated ✅

Added the future todo ✅